File: 160/intersect_two_llist.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        pA = headA
        pB = headB

        # compare sizes
        sizeA = self.get_size(pA)
        sizeB = self.get_size(pB)
        diff = abs(sizeA - sizeB)

        # traverse bigger list until equal
        p = pA
        if sizeB > sizeA:
        	p = pB
        while diff != 0:
        	p = p.next
        	diff -= 1
        if sizeB > sizeA:
        	pB = p
        else:
        	pA = p
        assert(self.get_size(p) == self.get_size(pA) or self.get_size(p) == self.get_size(pB))        

        if pA == None or pB == None:
        	return None

        print "nA: ", pA.val, "@ ", id(pA)
        print "nB: ", pB.val, "@ ", id(pB)

        # traverse in parallel
        while pA != None and pB != None:

        	if pA is pB:
        		return pA

        	pA = pA.next
        	pB = pB.next

        return None

    def get_size(self, ll_head):
    	node = ll_head
    	size = 0
    	while node != None:
    		size += 1
    		node = node.next
    	return size

File: 160/IntersectTwoLList.java

// Definition for singly-linked list.
class ListNode {
	int val;
	ListNode next;
	ListNode(int x) {
		val = x;
		next = null;
	}
}

public class IntersectTwoLList {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
    	int lenA = getLength(headA);
    	int lenB = getLength(headB);
    	ListNode pA = headA;
    	ListNode pB = headB;

    	// definitely no intersection
    	if (lenA == 0 || lenB == 0)
    		return null;

    	// start at same distance from interection
    	if (lenA < lenB) while (lenA < lenB) {pB = pB.next; lenB--;}
    	if (lenA > lenB) while (lenA > lenB) {pA = pA.next; lenA--;}

    	// move both pointers until match
    	while (lenA > 0) {
    		if (pA == pB)
    			return pA;
    		pA = pA.next;
    		pB = pB.next;
    		lenA--;
    		lenB--;
    	}
    	// no intersection
    	return null;
    }

    private int getLength(ListNode node) {
    	int len = 0;
    	while (node != null) {
    		len++;
    		node = node.next;
    	}
    	return len;
    }

    public static void main(String args[]) {
    	System.out.println("Hello World!");
    }
}